src/ParallelHashMap.h

```diff
@@ -0,0 +1,838 @@
+    K key;
+    V value;
+    node *next;
+  };
+
+  private:
+    size_t _M;      /* table size */
+    size_t _N;      /* number of elements present in table */
+    node ** _buckets;   /* buckets of values stored in nodes */
+
+  public:
+
+    fixed_hash_map(size_t M = 64);
+    virtual ~fixed_hash_map();
+    bool contains(K key);
+    V& at(K key);
+    void insert(K key, V value);
+    int insert_and_get_count(K key, V value);
+    size_t size();
+    size_t bucket_count();
+    K* keys();
+    V* values();
+    void clear();
+    void print_buckets();
+};
+
+/**
+ * @class ParallelHashMap ParallelHashMap.h "src/ParallelHashMap.h"
+ * @brief A thread-safe hash map supporting insertion and lookup operations
+ * @details The ParallelHashMap class is built ontop of the fixed_hash_map
+ *    class, supporting insertion and lookup operations but not deletion as
+ *    deletion is not needed in the OpenMOC application. This hash table uses
+ *    chaining for collisions, as defined in fixed_hash_map. It offers lock
+ *    free lookups in O(1) time on average and fine-grained locking for
+ *    insertions in O(1) time on average as well. Resizing is conducted
+ *    periodically during inserts, although the starting table size can be
+ *    chosen to limit the number of resizing operations.
+ */
+template <class K, class V>
+class ParallelHashMap
+{
+  /* padded pointer to hash table to avoid false sharing */
+  struct paddedPointer
+  {
+    volatile long pad_L1;
+    volatile long pad_L2;
+    volatile long pad_L3;
+    volatile long pad_L4;
+    volatile long pad_L5;
+    volatile long pad_L7;
+    volatile long pad_L8;
+    fixed_hash_map<K,V>* volatile value;
+    volatile long pad_R1;
+    volatile long pad_R2;
+    volatile long pad_R3;
+    volatile long pad_R4;
+    volatile long pad_R5;
+    volatile long pad_R6;
+    volatile long pad_R7;
+    volatile long pad_R8;
+  };
+  private:
+    fixed_hash_map<K,V> *_table;
+    paddedPointer *_announce;
+    size_t _num_threads;
+    size_t _N;
+    #ifdef OPENMP
+    omp_lock_t * _locks;
+    size_t _num_locks;
+    #endif
+    void resize();
+
+  public:
+    ParallelHashMap(size_t M = 64, size_t L = 64);
+    virtual ~ParallelHashMap();
+    bool contains(K key);
+    V& at(K key);
+    void insert(K key, V value);
+    int insert_and_get_count(K key, V value);
+    size_t size();
+    size_t bucket_count();
+    size_t num_locks();
+    K* keys();
+    V* values();
+    void clear();
+    void print_buckets();
+};
+
+/**
+ * @brief Constructor initializes fixed-size table of buckets filled with empty
+ *      linked lists.
+ * @details The constructor initializes a fixed-size hash map with the size
+ *      as an input parameter. If no size is given the default size (64)
+ *      is used. Buckets are filled with empty linked lists presented as
+ *      NULL pointers.
+ * @param M size of fixed hash map
+ */
+template <class K, class V>
+fixed_hash_map<K,V>::fixed_hash_map(size_t M)
```

Comment:
I am keeping a generic version of the hash table open source on my github. This hash table I want to conform to OpenMOC style.